fix(maxBy, minBy): return undefined for empty array inputs

umsungjun opened the pull request:
- Added an early return undefined when the input array is empty in both maxBy and minBy functions. This prevents the function from returning the initial element incorrectly when there are no items to evaluate.

- Updated the return type of maxBy to explicitly allow undefined. This reflects the correct behavior when the input array has no elements.

**Benchmarks**
![스크린샷 2025-04-17 오후 9 09 47](https://github.com/user-attachments/assets/8141a81d-1d63-4d6f-96d3-01c2259c7ab4)

hsskey commented:
Nice fix! Just a small suggestion: since this PR changes the behavior for empty arrays, can you add a test case for that? It'll help prevent regressions in the future and document the expected behavior more clearly. Thanks 🙌

umsungjun commented:
To. @hsskey 

Thanks for the suggestion! 🙌
The test case for the empty array scenario was actually already present in minBy.spec.ts and maxBy.spec.ts.
Let me know if there’s anything else I should adjust!

<img width="878" alt="스크린샷 2025-04-18 오후 2 53 33" src="https://github.com/user-attachments/assets/72392895-8030-4ed8-b3c7-3d860e4dad18" />

hsskey commented:
Thanks for showing me! I see the tests were already there.

hsskey commented:
> To. @hsskey
> 
> Thanks for the suggestion! 🙌 The test case for the empty array scenario was actually already present in minBy.spec.ts and maxBy.spec.ts. Let me know if there’s anything else I should adjust!
> 
> <img alt="스크린샷 2025-04-18 오후 2 53 33" width="878" src="https://private-user-images.githubusercontent.com/99730910/435078753-72392895-8030-4ed8-b3c7-3d860e4dad18.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ5NTY4OTksIm5iZiI6MTc0NDk1NjU5OSwicGF0aCI6Ii85OTczMDkxMC80MzUwNzg3NTMtNzIzOTI4OTUtODAzMC00ZWQ4LWIzYzctM2Q4NjBlNGRhZDE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE4VDA2MDk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjODdiN2MzYTJhMzJhMDA1ZWY0ZDFkNGQzYTJmNWMxYjA2ZWU3ZWNjNjgxMzViNDkzYjQ0OGY2ZjM3ODI2YjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qeLuFDbbijnQH6pX4kR_T6MA5o_jWyaKH7e7RuNAuUs">

Thanks for showing me! I see the tests were already there.

raon0211 commented:
Hey! Thanks for opening the pull request. I appreciate the effort, but I believe the current behavior is actually correct.

In JavaScript, `Math.min()` returns `Infinity` when called with no arguments, and `Math.max()` returns `-Infinity`. This might seem odd at first, but it makes sense when you think about how these functions work.

For example, `Math.min(1, 2, 3)` will always be greater than or equal to `Math.min(1, 2, 3, x)` for any number `x`. As you add more values, the result can only stay the same or get smaller.

Now, take that idea to the extreme: if you call `Math.min()` with no values at all, the result should be greater than or equal to `Math.min(x)` for any `x`. That means it must return the largest possible value — which is `Infinity`.

umsungjun commented:
Thank you for the feedback! @raon0211 

I’ve updated both maxBy and minBy to return undefined when the input array is empty.
The reason for this change is that the test cases expect undefined when the array is empty, so I added the condition to match that expectation.

<img width="1177" alt="스크린샷 2025-04-19 오후 10 51 30" src="https://github.com/user-attachments/assets/df24062b-ecfb-4147-90ac-149cad982d0a" />


I understand your explanation about Math.min() and Math.max() returning Infinity or -Infinity when called without arguments.
However, in this case, since minBy and maxBy are designed to return an element from the array, I believe returning undefined when the array is empty is the correct behavior.

If I’ve misunderstood anything, please let me know!

umsungjun commented:
Thank you for the feedback!

I understand now.
```tsx
const emptyArray = [];
console.log(emptyArray[Infinity]); // undefined
```
It returns undefined.

raon0211 commented:
Thanks for your understanding :)
